REGRESSION(290778@main): [Unified PDF] [iPadOS] Content offset lost when returning from another tab

aprotyas opened the pull request:
#### b8823e2624731256d41693403c9d57599564055d
<pre>
REGRESSION(290778@main): [Unified PDF] [iPadOS] Content offset lost when returning from another tab
<a href="https://bugs.webkit.org/show_bug.cgi?id=291910">https://bugs.webkit.org/show_bug.cgi?id=291910</a>
<a href="https://rdar.apple.com/148822838">rdar://148822838</a>

Reviewed by Tim Horton.

On iPadOS, tab switching in Safari kicks off a view hierarchy layout
update that ends up producing an override of the active WKWebView&apos;s
layout parameters with an empty layout size. This empty layout size
leads to resizing the PluginWidget with an empty frame rect, which
messes up the internal plugin layout, and results in an apparent loss of
scroll offset.

This patch rejects such updates for main frame plugins on iOS. I limit
these rejections to main frame plugins because there may really be a use
for these empty viewport updates for general web content that I do not
understand yet.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setViewportConfigurationViewLayoutSize):

Reject viewport configuration updates if an empty view layout size is
served when mainFramePlugInRejectsZeroSizeViewLayoutSizeUpdates == YES.

(WebKit::WebPage::mainFramePlugInRejectsZeroViewLayoutSizeUpdates const):

Codify the new behavior for main frame plugins on iOS.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

API test: UnifiedPDF.ScrollOffsetUnchangedWithZeroSizeViewportUpdate

This test forces an intermediate zero-sized viewport update and verifies
that the scroll offset stays unchanged.

Canonical link: <a href="https://commits.webkit.org/293983@main">https://commits.webkit.org/293983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9702953a9c0bcd60a5e25c5ab99a49333d761ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20108 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10407 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105593 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51044 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76493 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33546 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103463 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15644 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56848 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15461 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8746 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50420 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107947 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27574 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20240 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85445 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27937 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86943 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84983 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21624 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29668 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7402 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21539 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27509 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32759 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27320 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30638 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28878 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 reviewed (approved):
This seems like a bit of a hack around a deeper problem, but if it gets us out of the woods for PDF, and the deeper problem doesn't affect web content, it seems like a fine workaround.

aprotyas committed:
REGRESSION(290778@main): [Unified PDF] [iPadOS] Content offset lost when returning from another tab
https://bugs.webkit.org/show_bug.cgi?id=291910
rdar://148822838

Reviewed by Tim Horton.

On iPadOS, tab switching in Safari kicks off a view hierarchy layout
update that ends up producing an override of the active WKWebView's
layout parameters with an empty layout size. This empty layout size
leads to resizing the PluginWidget with an empty frame rect, which
messes up the internal plugin layout, and results in an apparent loss of
scroll offset.

This patch rejects such updates for main frame plugins on iOS. I limit
these rejections to main frame plugins because there may really be a use
for these empty viewport updates for general web content that I do not
understand yet.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setViewportConfigurationViewLayoutSize):

Reject viewport configuration updates if an empty view layout size is
served when mainFramePlugInRejectsZeroSizeViewLayoutSizeUpdates == YES.

(WebKit::WebPage::mainFramePlugInRejectsZeroViewLayoutSizeUpdates const):

Codify the new behavior for main frame plugins on iOS.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

API test: UnifiedPDF.ScrollOffsetUnchangedWithZeroSizeViewportUpdate

This test forces an intermediate zero-sized viewport update and verifies
that the scroll offset stays unchanged.

Canonical link: https://commits.webkit.org/293983@main
